src/signatures/ternary/wots.rs

```diff
@@ -0,0 +1,367 @@
+    // content can't go wrong.
+    Ok(unsafe {
+        Trits::<T3B1>::from_raw_unchecked(&normalized, normalized.len() * 3)
+            .to_buf::<T3B1Buf>()
+            .encode::<T1B1Buf>()
+    })
+}
+
+/// Winternitz One Time Signature private key.
+#[derive(SecretDebug, SecretDisplay, SecretDrop)]
+pub struct WotsPrivateKey<S> {
+    pub(crate) state: TritBuf<T1B1Buf>,
+    pub(crate) marker: PhantomData<S>,
+}
+
+impl<S> Zeroize for WotsPrivateKey<S> {
+    fn zeroize(&mut self) {
+        // This unsafe is fine since we only reset the whole buffer with zeros, there is no alignement issues.
+        unsafe { self.state.as_i8_slice_mut().zeroize() }
+    }
+}
+
+impl<S: Sponge + Default> PrivateKey for WotsPrivateKey<S> {
+    type PublicKey = WotsPublicKey<S>;
+    type Signature = WotsSignature<S>;
+    type Error = Error;
+
+    fn generate_public_key(&self) -> Result<Self::PublicKey, Self::Error> {
+        let mut sponge = S::default();
+        let mut hashed_private_key = self.state.clone();
+        let security = self.state.len() / SIGNATURE_FRAGMENT_LENGTH;
+        let mut digests = TritBuf::<T1B1Buf>::zeros(security * HASH_LENGTH);
+        let mut public_key_state = TritBuf::<T1B1Buf>::zeros(HASH_LENGTH);
+
+        // Hash each chunk of the private key the maximum amount of times.
+        for chunk in hashed_private_key.chunks_mut(HASH_LENGTH) {
+            for _ in 0..Tryte::MAX_VALUE as i8 - Tryte::MIN_VALUE as i8 {
+                sponge
+                    .absorb(chunk)
+                    .and_then(|_| sponge.squeeze_into(chunk))
+                    .map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge.reset();
+            }
+        }
+
+        // Create one digest per fragment of the private key.
+        for (i, chunk) in hashed_private_key.chunks(SIGNATURE_FRAGMENT_LENGTH).enumerate() {
+            sponge
+                .digest_into(chunk, &mut digests[i * HASH_LENGTH..(i + 1) * HASH_LENGTH])
+                .map_err(|_| Self::Error::FailedSpongeOperation)?;
+        }
+
+        // Hash the digests together to produce the public key.
+        sponge
+            .digest_into(&digests, &mut public_key_state)
+            .map_err(|_| Self::Error::FailedSpongeOperation)?;
+
+        Ok(Self::PublicKey {
+            state: public_key_state,
+            marker: PhantomData,
+        })
+    }
+
+    fn sign(&mut self, message: &Trits<T1B1>) -> Result<Self::Signature, Self::Error> {
+        if message.len() != HASH_LENGTH {
+            return Err(Error::InvalidMessageLength(message.len()));
+        }
+
+        let mut sponge = S::default();
+        let mut signature = self.state.clone();
+
+        for (f, fragment) in signature.chunks_mut(SIGNATURE_FRAGMENT_LENGTH).enumerate() {
+            let message_fragment =
+                &message[f % 3 * MESSAGE_FRAGMENT_LENGTH * 3..(f % 3 + 1) * MESSAGE_FRAGMENT_LENGTH * 3];
```

Comment:
The `3` here actually comes from 
```
MESSAGE_FRAGMENT_LENGTH_TRITS = MESSAGE_FRAGMENT_LENGTH(_TRYTES) * 3
HASH_LENGTH / MESSAGE_FRAGMENT_LENGTH_TRITS
```
Maybe using a definition using the consts is a bit clearer than a magic number especially with the unwrap a few lines down